# Introduction to Next Generation Sequencing data
## A course of the Summer School of Aarhus University
**Computing and didactical support from the [Danish Healt Data Science Sandbox](https://hds-sandbox.github.io/)**

The material for this course is organized in four separated jupyter notebooks in both `bash`, `python` and `R` where you will benefit of an interactive coding setup on `jupyterlab`. 



<!-- OVERVIEW OF COURSE -->
!!! abstract "Overview"

    After the course, you will have knowledge of bioinformatics methods for analyzing genomes using NGS data, including knowledge of the existing types of genome data, how the different types of data can be displayed and analyzed, the current methods for genome assembly and analysis, their accuracy and how they can be used. The course will enable you to devise and run a project that makes use of NGS data.

    &#128218; **Prerequisites:** This is an introductory course that needs a basic understanding of the biology behind sequencing, and not necessarily programming experience (though this would help!).


    &#128172; **Syllabus:**     
        1. Describe key challenges in the analysis of NGS data <br>
        2. Explain the theoretical foundation for methods that use NGS for assembly and analysis of genomes<br>
        3. Discuss the bioinformatic methods for genome analysis and hypothesize what drives the outcome of the methods<br>
        4. Review original literature within the subjects and relate the discussed topics to analysis scenarios<br>
        5. Apply bioinformatics tools within the selected application areas and reflect on the results, formulating your own conclusion in the proposed tasks<br>

    &#128368; **Total Time Estimation:** 20 hours  

    &#128193; **Supporting Materials:**  

    - jupyter notebooks for interactive coding
    - lecture slides from the instructor

    You can find the links to the material in the table at the bottom of this page.

    &#128397; **Course authors and instructors:**

    [Mikkel H Schierup](https://pure.au.dk/portal/da/persons/mikkel-heide-schierup(aa7cd7f5-b8cf-4781-b6a6-be522a7b335c).html){ .md-button }
    [Stig U Andersen](https://pure.au.dk/portal/da/persons/stig-uggerhoej-andersen(dbcd376b-110b-489a-8f6d-85dd2268277d).html){ .md-button }

    [Samuele Soraggi](https://www.linkedin.com/in/samuelesoraggi/){ .md-button }
    [Peter S Porsborg](https://www.linkedin.com/in/peter-s%C3%B8rud-porsborg/){ .md-button }
    [Adrián G Repollés](https://www.linkedin.com/in/adri%C3%A1n-g%C3%B3mez-repoll%C3%A9s-66b496105/?originalSubdomain=es){ .md-button }

    &#128203; **License:** [Tutorial Content is licensed under Creative Commons Attribution 4.0 International License](https://creativecommons.org/licenses/by/4.0/)

    &#128221; **Citation:** If you use any of this material for your research, please cite this course with the DOI below, and acknowledge the Health Data Science Sandbox project of the Novo Nordisk Foundation (grant number NNF20OC0063268). It is of great help to support the project.
    [![DOI](https://zenodo.org/badge/468293635.svg)](https://zenodo.org/badge/latestdoi/468293635)

    &#128231; **Contact:** Samuele Soraggi (samuele at birc.au.dk).


!!! warning "Course structure and Instructions"

    The course exercises are organized in four exercise modules. 
    
    The first one is executed with the web interface [usegalaxy.org](https://usegalaxy.org/). Click on `1.Galaxy Exercise` in the menu for instructions)
    
    Afterwards, we will work on a computing environment to use `jupyterlab`. Use `2.Instructions` for instructions.

    The menu `3.Course exercises` contains all the compiled exercises as a reference.


## Course material 2022

Here you find a table with the instructor's slides from 2022.

| Topic  | Slide | Notebook |
| ------------- | ------------- | ------------- |
| Sequencing technologies  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/01_sequencing_technologies.pdf)  | -- |
| Mapping to reference  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/02_mapping_to_reference.pdf)  | [Notebook](https://hds-sandbox.github.io/NGS_summer_course_Aarhus/nb/align.html) |
| Data visualization  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/03_data_visualization.pdf)  | -- |
| SNPs and structural variants  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/04_snp_and_structural_variants.pdf)  | [Notebook](https://hds-sandbox.github.io/NGS_summer_course_Aarhus/nb/vcf.html) |
| RNA sequencing  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/05_rna_sequencing.pdf)  | [Notebook](https://hds-sandbox.github.io/NGS_summer_course_Aarhus/nb/bulk.html) |
| De-novo assembly  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/06_de_novo_assembly.pdf)  | -- |
| Microbiomes and metagenomics  | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/07_microbiomes_and_metagenomics.pdf)  | -- |
| Single cell RNA sequencing | [link](https://github.com/hds-sandbox/NGS_summer_course_Aarhus/blob/42d3a6cad8e64b47dac01bc2995927fa4c39333d/Slides/08_single_cell_day.pdf)  | [Notebook](https://hds-sandbox.github.io/NGS_summer_course_Aarhus/nb/scrna.html) |

## Course material 2023 (on its way after the course's end)

Here you find a table with the instructor's slides and a link to the compiled notebooks, that you can also run on your own following the `instructions` in this webpage. Data alignment can also be performed on the `Galaxy` interactive webpage (see the `galaxy exercise` in this webpage).